examples/counter/shared/src/app/mod.rs

```diff
@@ -0,0 +1,172 @@
+    type Model = Model;
+    type Event = Event;
+    type ViewModel = ViewModel;
+    type Capabilities = Capabilities;
+
+    fn update(&self, msg: Self::Event, model: &mut Self::Model, caps: &Self::Capabilities) {
+        match msg {
+            Event::Get => {
+                caps.http.get_json(Url::parse(API_URL).unwrap(), Event::Set);
+            }
+            Event::Set(counter) => {
+                model.count = counter.value;
+                caps.render.render();
+            }
+            Event::Increment => {
+                let base = Url::parse(API_URL).unwrap();
+                let url = base.join("/inc");
+                caps.http.post(url.unwrap(), Event::Set)
+            }
+            Event::Decrement => {
+                let base = Url::parse(API_URL).unwrap();
+                let url = base.join("/dec");
+                caps.http.post(url.unwrap(), Event::Set)
+            }
+        }
+    }
+
+    fn view(&self, model: &Self::Model) -> Self::ViewModel {
+        model.into()
+    }
+}
+
+#[derive(Default)]
+pub struct Model {
+    count: isize,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct ViewModel {
+    pub count: isize,
+}
+
+impl From<&Model> for ViewModel {
+    fn from(model: &Model) -> Self {
+        Self { count: model.count }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub enum Event {
+    Get,
+    Set(Counter),
+    Increment,
+    Decrement,
+}
+
+pub struct Capabilities {
+    pub http: Http<Event>,
+    pub render: Render<Event>,
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub struct Counter {
+    value: isize,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{App, Event, Model};
+    use crate::{Counter, Effect};
+    use crux_core::testing::AppTester;
+    use crux_http::{HttpRequest, HttpResponse};
+
+    #[test]
+    fn get_counter() {
+        let app = AppTester::<App, _>::default();
+        let mut model = Model::default();
+
+        let update = app.update(Event::Get, &mut model);
+
+        let actual = &update.effects[0];
+        let expected = &Effect::Http(HttpRequest {
+            method: "GET".to_string(),
+            url: "https://crux-counter.fly.dev/".to_string(),
+        });
+        assert_eq!(actual, expected);
+
+        let update = update.effects[0].resolve(&HttpResponse {
+            status: 200,
+            body: serde_json::to_vec(&Counter { value: 1 }).unwrap(),
+        });
+
+        let actual = update.events;
+        let expected = vec![Event::Set(Counter { value: 1 })];
+        assert_eq!(actual, expected);
+    }
+
+    #[test]
+    fn set_counter() {
+        let app = AppTester::<App, _>::default();
+        let mut model = Model::default();
+
+        let update = app.update(Event::Set(Counter { value: 1 }), &mut model);
+
+        let actual = &update.effects[0];
+        let expected = &Effect::Render;
+        assert_eq!(actual, expected);
+
+        let actual = model.count;
```

Comment:
should we also test `view()` here?

Reply:
I would, possibly before checking the model. In some sense it's black box (via view) vs white box (via model) testing.

Reply:
Yeah we'll need to expose it

Reply:
will do this in another pr